src/Features/CSharp/Portable/ConvertForToForEach/CSharpConvertForToForEachCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,163 @@
+using Microsoft.CodeAnalysis.Options;
+using Microsoft.CodeAnalysis.Text;
+
+namespace Microsoft.CodeAnalysis.CSharp.ConvertForToForEach
+{
+    [ExportCodeRefactoringProvider(LanguageNames.CSharp, Name = nameof(CSharpConvertForToForEachCodeRefactoringProvider)), Shared]
+    internal class CSharpConvertForToForEachCodeRefactoringProvider :
+        AbstractConvertForToForEachCodeRefactoringProvider<
+            StatementSyntax, 
+            ForStatementSyntax, 
+            ExpressionSyntax,
+            MemberAccessExpressionSyntax,
+            TypeSyntax,
+            VariableDeclaratorSyntax>
+    {
+        protected override string GetTitle()
+            => CSharpFeaturesResources.Convert_for_to_foreach;
+
+        protected override bool IsValidCursorPosition(ForStatementSyntax forStatement, int cursorPos)
+        {
+            // If there isn't a selection, then we allow the refactoring from the start of
+            // 'for' to the start of the open paren, or in the trailing trivia of the c
+            // close paren.
+            var startSpan = TextSpan.FromBounds(forStatement.ForKeyword.SpanStart, forStatement.OpenParenToken.SpanStart);
+            var endSpan = TextSpan.FromBounds(forStatement.CloseParenToken.Span.End, forStatement.CloseParenToken.FullSpan.End);
+
+            return startSpan.IntersectsWith(cursorPos) || endSpan.IntersectsWith(cursorPos);
+        }
+
+        protected override SyntaxList<StatementSyntax> GetBodyStatements(ForStatementSyntax forStatement)
+            => forStatement.Statement is BlockSyntax block
+                ? block.Statements
+                : SyntaxFactory.SingletonList(forStatement.Statement);
+
+        protected override bool TryGetForStatementComponents(
+            ForStatementSyntax forStatement,
+            out SyntaxToken iterationVariable, out ExpressionSyntax initializer,
+            out MemberAccessExpressionSyntax memberAccess, 
+            out ExpressionSyntax stepValueExpressionOpt,
+            CancellationToken cancellationToken)
+        {
+            // Look for very specific forms.  Basically, only minor variations around:
+            // for (var i = 0; i < expr.Lenth; i++)
+
+            if (forStatement.Declaration != null &&
+                forStatement.Condition.IsKind(SyntaxKind.LessThanExpression) &&
+                forStatement.Incrementors.Count == 1)
+            {
+                var declaration = forStatement.Declaration;
+                if (declaration.Variables.Count == 1)
+                {
+                    var declarator = declaration.Variables[0];
+                    if (declarator.Initializer != null)
+                    {
+                        iterationVariable = declarator.Identifier;
+                        initializer = declarator.Initializer.Value;
+
+                        var binaryExpression = (BinaryExpressionSyntax)forStatement.Condition;
+
+                        // Look for:  i < expr.Length
```

Comment:
Do we bail out if `<` operator is user defined?

Reply:
No.  This is a pattern based refactoring.  This is not an analyzer+fix.  The goal is to allow the user to easily move code that has such patterns over to using foreach.